clippy_lints/src/if_then_some_else_none.rs

```diff
@@ -77,7 +80,18 @@ impl<'tcx> LateLintPass<'tcx> for IfThenSomeElseNone {
             && !is_in_const_context(cx)
             && self.msrv.meets(cx, msrvs::BOOL_THEN)
             && !contains_return(then_block.stmts)
-            && then_block.expr.is_none_or(|expr| !contains_return(expr))
+            && then_block.expr.is_none_or(|expr| {
+                !contains_return(expr)
+                    // `bool::then` doesn't work with async code
+                    && for_each_expr_without_closures(expr, |e| {
+                        if desugar_await(e).is_some() {
+                            ControlFlow::Break(())
+                        } else {
+                            ControlFlow::Continue(())
+                        }
+                    })
+                    .is_none()
+            })
```

Comment:
Could we just use [`can_move_expr_to_closure`](https://doc.rust-lang.org/nightly/nightly-rustc/clippy_utils/fn.can_move_expr_to_closure.html) on the `then` expression here? I'm not sure if it would introduce FNs but it seems like it answers exactly the question of "can we move that code into a closure", taking into account control flow constructs and the like.